src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,414 @@
+ */
+#ifndef MLPACK_METHODS_RL_ENVIRONMENT_FROZEN_LAKE_HPP
+#define MLPACK_METHODS_RL_ENVIRONMENT_FROZEN_LAKE_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <stack>
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of Cart Pole task.
+ */
+class FrozenLake
+{
+ public:
+  class State
+  {
+   public:
+    State()
+    { 
+      // Nothing to do here.
+    }
+
+    // Initialize a new state with current row and column at 0.
+    State(size_t nRows, size_t nCols) : 
+      nRows(nRows),
+      nCols(nCols),
+      curRow(0),
+      curCol(0)
+    {
+      // Nothing to do here.
+    }
+
+    //! Copy constructor.
+    State(State const& other) :
+      nRows(other.nRows),
+      nCols(other.nCols),
+      curRow(other.curRow),
+      curCol(other.curCol)
+    {
+      // Nothing to do here.
+    }
+
+    //! Move constructor.
+    State(State&& other) :
+      nRows(other.nRows),
+      nCols(other.nCols),
+      curRow(other.curRow),
+      curCol(other.curCol)
+    {
+      // Nothing to do here.
+    }
+
+    //! Operator= copy constructor.
+    State& operator=(State const& other)
+    {
+      if (this != &other)
+      {
+        nRows = other.nRows;
+        nCols = other.nCols;
+        curRow = other.curRow;
+        curCol = other.curCol;
+      }
+      return *this;
+    }
+
+    //! Operator= move constructor.
+    State& operator=(State&& other)
+    {
+      if (this != &other)
+      {
+        nRows = other.nRows;
+        nCols = other.nCols;
+        curRow = other.curRow;
+        curCol = other.curCol;
+      }
+      return *this;
+    }
+
+    //! Get nRows.
+    size_t NumRows() const {return nRows;}
+
+    //! Get nCol.
+    size_t NumCols() const {return nCols;}
+
+    //! Get current row.
+    size_t CurRow() const {return curRow;}
+
+    //! Get current col.
+    size_t CurCol() const {return curCol;}
+
+    //! Set current row.
+    size_t& CurRow() {return curRow;}
+
+    //! Set current col.
+    size_t& CurCol() {return curCol;}
+
+    //! Get state representation.
+    size_t GetState() const {return nCols * curRow + curCol;}
+
+   private:
+    size_t nRows;
+    size_t nCols;
+    size_t curRow;
+    size_t curCol;
+  };
+
+  /**
+   * Implementation of action of Frozen Lake.
+   */
+  class Action
+  {
+   public:
+    enum actions
+    {
+      Up, Down, Left, Right
+    };
+    // To store the action.
+    Action::actions action;
+
+    // Track the size of the action space.
+    static const size_t size = 4;
+  };
+
+  //! Constructor.
+  FrozenLake(size_t const& maxSteps=200,
+              size_t const& nRows=4,
+              size_t const& nCols=4,
+              double const& platformRate=0.8) : 
+    maxSteps(maxSteps),
+    nRows(nRows),
+    nCols(nCols)
+  { 
+    // Preprocess platform rate
+    this->platformRate = std::min(platformRate, 1.0);
+    this->platformRate = std::max(platformRate, 0.0);
+  }
+
+  //  Step function
```

Comment:
```suggestion
  // Step function
```

Remove some extra whitespace.